sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/OrcReadBenchmark.scala

```diff
@@ -0,0 +1,517 @@
+
+        String with Nulls Scan:                  Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
+        ------------------------------------------------------------------------------------------------
+        Native ORC MR (0.95%)                         1287 / 1331          8.2         122.7       1.0X
+        Native ORC Vectorized (0.95%)                  404 /  407         26.0          38.5       3.2X
+        Native ORC Vectorized (0.95%) (Java)           405 /  409         25.9          38.6       3.2X
+        Hive built-in ORC (0.95%)                     1612 / 1644          6.5         153.7       0.8X
+        */
+        benchmark.run()
+      }
+    }
+  }
+
+  def columnsBenchmark(values: Int, width: Int): Unit = {
+    val sqlBenchmark = new Benchmark(s"SQL Single Column Scan from wide table ($width columns)", values)
+
+    withTempPath { dir =>
+      withTempTable("t1", "nativeOrcTable", "hiveOrcTable") {
+        import spark.implicits._
+        val middle = width / 2
+        val selectExpr = (1 to width).map(i => s"value as c$i")
+        spark.range(values).map(_ => Random.nextLong).toDF()
+          .selectExpr(selectExpr: _*).createOrReplaceTempView("t1")
+
+        prepareTable(dir, spark.sql("SELECT * FROM t1"))
+
+        sqlBenchmark.addCase("Native ORC MR") { _ =>
+          withSQLConf(SQLConf.ORC_VECTORIZED_READER_ENABLED.key -> "false") {
+            spark.sql(s"SELECT sum(c$middle) FROM nativeOrcTable").collect()
+          }
+        }
+
+        sqlBenchmark.addCase("Native ORC Vectorized") { _ =>
+          withSQLConf(SQLConf.ORC_VECTORIZED_JAVA_READER_ENABLED.key -> "false") {
+            spark.sql(s"SELECT sum(c$middle) FROM nativeOrcTable").collect()
+          }
+        }
+
+        sqlBenchmark.addCase("Native ORC Vectorized (Java)") { _ =>
+          withSQLConf(SQLConf.ORC_VECTORIZED_JAVA_READER_ENABLED.key -> "true") {
+            spark.sql(s"SELECT sum(c$middle) FROM nativeOrcTable").collect()
+          }
+        }
+
+        sqlBenchmark.addCase("Hive built-in ORC") { _ =>
+          spark.sql(s"SELECT sum(c$middle) FROM hiveOrcTable").collect()
+        }
+
+        /*
+        Java HotSpot(TM) 64-Bit Server VM 1.8.0_152-b16 on Mac OS X 10.13.1
+        Intel(R) Core(TM) i7-4770HQ CPU @ 2.20GHz
+
+        SQL Single Column Scan from wide table (100 columns): Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
+        ------------------------------------------------------------------------------------------------
+        Native ORC MR                                 1116 / 1117          0.9        1064.7       1.0X
+        Native ORC Vectorized                           92 /   98         11.3          88.2      12.1X
+        Native ORC Vectorized (Java)                    90 /   96         11.6          86.0      12.4X
+        Hive built-in ORC                              376 /  386          2.8         358.6       3.0X
```

Comment:
hmm, why is hive ORC much faster than the native ORC for this case?

Reply:
It's because it has ORC MR reader now.
At my initial PR, I provided 3 readers.
1. ORC MR readers
2. ORC Vectorized Reader
3. ORC Vectorized Reader + Spark Columnar Batch Reader.

Due to the delaying, I only put (1) previouly for review. MR reader joins the columns. So, it's slow specifically in this case.

Reply:
do you mean Hive ORC reader is vectorized under the hood, but return data as rows to Spark?

Reply:
Ur, I meant only current ORC readers. I didn't check the detail of old Hive ORC reader.